app/scripts/lib/createMetaRPCHandler.js

```diff
@@ -1,6 +1,38 @@
+import log from 'loglevel';
 import { ethErrors, serializeError } from 'eth-rpc-errors';
+import { captureException } from '@sentry/browser';
+import ReadOnlyNetworkStore from './network-store';
+import LocalStore from './local-store';
 
-const createMetaRPCHandler = (api, outStream) => {
+let dataPersistenceFailing = false;
+
+async function persistData(state, localStore) {
```

Comment:
Suggest renaming this method to signify that it is a failsafe, fallback state persistence layer, and document it with jsdocs explaining the goal is to overwrite the state with a more up-to-date copy just before the disconnect happens. I don't want future reviewers of this code to believe that this is the primary persistence layer.

Reply:
Well actually it kindof is the primary persistence layer, as it will always get called before the persist call [here](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/background.js#L357). As we discussed in slack, I'm not sure whether we ought to even keep the other persist call (linked in the previous sentence) any longer., since I believe it will produce redundant persists as it currently stands in this PR. @Gudahtt hoping you could chime in here.

Reply:
removed the old persistence layer [here](https://github.com/MetaMask/metamask-extension/pull/15978/commits/90f5e60903a092fc9b9163b06078efe3a44915e0) (still needs some cleanup)... not sure if this is for keeps but gave it a whirl and behaves correctly.

Reply:
Hmm, why would the old persistence layer be removed? Not sure I follow. I thought we still wanted to ensure state was persisted even if there was no RPC traffic. For example, if a transaction settles while the user is not using MetaMask.

Reply:
I had thought that this would pass through the RPC client too, but I guess it doesn't... So are we ok with all the redundant calls that will happen with both layers? I suppose we have to live with it for now?

Reply:
I do not think we have to just live with it, no. We can prevent redundant calls by writing our own `debounce` logic to remove this duplication.

Reply:
hmmm right okay. I think I was kindof confused by the use of debounce, because its sortof a different concept right? Just call persist every second and batch any requests that came during that second.

Reply:
> The first condition is upheld right now just by using debounce, but we can scrap that and implement our own custom logic to accomplish both.

As part of this are you suggesting that we somehow identify if the state update has already been accomplished by the action taken to
> Persist state immediately before returning any JSON-RPC response.
?

Sorry this is getting a bit convoluted, but basically it seems if we keep both `persistData` calls as they stand we're achieving both conditions, we're just getting a bunch of "redundant" calls because each time we persist the immediately before returning any JSON-RPC response,  we also then `persistData` again [here](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/background.js#L357)?

Reply:
Yes, you're right, with the redundant calls we accomplish both goals. I was hoping we could avoid the redundant calls but we don't need that to be done here. We can try that later.